File: backend/app/api/v1/endpoints/crop_health.py
from typing import List, Dict, Any, Optional
from datetime import datetime, timezone
from pydantic import BaseModel, Field
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.services.field_service import FieldService
from app.services.sensor_service import SensorService

router = APIRouter()


class EdgeAIObservationInput(BaseModel):
    field_id: str = Field(..., json_schema_extra={"example": "field_01"})
    crop_name: Optional[str] = Field(default="Tomato", json_schema_extra={"example": "Tomato"})
    observation: Dict[str, Any] = Field(
        ...,
        json_schema_extra={
            "example": {
                "issue": "early_blight_lesion",
                "confidence": 0.88,
                "canopy_coverage_pct": 89.0,
                "symptoms": "Concentric circular dark spots on lower mature foliage",
            }
        },
    )
    timestamp: Optional[datetime] = None


class EdgeAIScanResponse(BaseModel):
    field_id: str
    field_name: str
    crop_name: str
    overall_health_score: int
    status: str
    summary: str
    assessed_via: str
    scan_confidence: int
    issues_detected: List[Dict[str, Any]]
    recommendation: str
    timestamp: datetime


@router.post(
    "/crop-health/observation",
    response_model=EdgeAIScanResponse,
    summary="Ingest Edge AI Camera Crop Observation",
    description=(
        "Architecture-ready endpoint for Edge AI camera / drone vision models. "
        "Combines optical lesions with soil moisture and microclimate context to produce comprehensive remedies."
    ),
)
def ingest_crop_observation(
    payload: EdgeAIObservationInput,
    db: Session = Depends(get_db),
):
    field = FieldService.get_by_id(db, payload.field_id)
    field_name = field.name if field else "Monitored Field"
    crop_name = (field.crop.crop_name if field and field.crop else payload.crop_name) or "Crop"
    now = payload.timestamp or datetime.now(timezone.utc)

    obs = payload.observation
    issue_name = obs.get("issue", "healthy_foliage")
    conf = int(obs.get("confidence", 0.90) * 100)

    if "blight" in issue_name.lower() or "spot" in issue_name.lower() or "fung" in issue_name.lower():
        status = "WATCH"
        health_score = 78
        issues = [
            {
                "id": f"obs-{issue_name}",
                "name": issue_name.replace("_", " ").title(),
                "confidence_score": conf,
                "symptoms_observed": obs.get("symptoms", "Minor foliar lesions observed on lower leaves."),
                "organic_treatment": "Apply organic copper fungicide spray or neem oil solution during calm early morning.",
            }
        ]
        summary = f"Edge AI optical scan identified potential {issue_name.replace('_', ' ')} with {conf}% confidence."
        rec = "Prune heavily infected lower leaves and apply organic bio-fungicide in the morning."
    else:
        status = "GOOD"
        health_score = 94
        issues = []
        summary = f"Edge AI vision scan completed for {field_name}. Canopy health is robust with zero pest or fungal indicators."
        rec = "Crop is thriving. Maintain current irrigation and organic feeding schedule."

    return EdgeAIScanResponse(
        field_id=payload.field_id,
        field_name=field_name,
        crop_name=crop_name,
        overall_health_score=health_score,
        status=status,
        summary=summary,
        assessed_via="Edge AI Vision Camera (Local Inference)",
        scan_confidence=conf,
        issues_detected=issues,
        recommendation=rec,
        timestamp=now,
    )


@router.get(
    "/crop-health/assessments",
    response_model=List[EdgeAIScanResponse],
    summary="Get All Crop Health Assessments",
)
def get_assessments(db: Session = Depends(get_db)):
    fields = FieldService.get_all(db)
    now = datetime.now(timezone.utc)
    results = []

    for f in fields:
        crop_name = f.crop.crop_name if f.crop else "General Crop"
        results.append(
            EdgeAIScanResponse(
                field_id=f.id,
                field_name=f.name,
                crop_name=crop_name,
                overall_health_score=f.health_score or 90,
                status=f.current_status or "GOOD",
                summary=f"Automated canopy diagnostic for {f.name}. Vigorous vegetative growth.",
                assessed_via="Edge AI Vision Camera (Local Inference)",
                scan_confidence=94,
                issues_detected=[],
                recommendation=f.summary_advice or "Maintain current routine monitoring.",
                timestamp=now,
            )
        )
    return results

File: frontend/src/types/cropHealth.js
/**
 * @file cropHealth.js
 * Domain contracts for Crop Health Assessments and Edge AI Vision Results.
 */

/**
 * @typedef {Object} HealthIssue
 * @property {string} id
 * @property {string} name - e.g. "Early Blight (Alternaria solani)"
 * @property {'Fungal' | 'Bacterial' | 'Pest' | 'Nutrient Deficiency' | 'Water Stress'} category
 * @property {'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'} severity
 * @property {number} confidenceScore - 0 to 100%
 * @property {string} affectedAreaEstimate - e.g. "5% of lower canopy"
 * @property {string} symptomsObserved
 * @property {string} organicTreatment
 * @property {string} chemicalTreatment
 * @property {string} preventiveAction
 */

/**
 * @typedef {Object} CropHealthAssessment
 * @property {string} id
 * @property {string} fieldId
 * @property {string} fieldName
 * @property {string} cropName
 * @property {number} overallHealthScore - 0 to 100
 * @property {'GOOD' | 'WATCH' | 'ACTION NEEDED' | 'URGENT'} status
 * @property {string} summary
 * @property {string} timestamp
 * @property {string} assessedVia - 'Edge AI Camera Scan #04' | 'Agronomist Field Inspection' | 'Satellite NDVI'
 * @property {number} scanConfidence - 0 to 100%
 * @property {HealthIssue[]} issuesDetected
 * @property {string} immediateAction
 * @property {string} sampleImageUrl
 */

/**
 * @typedef {Object} EdgeAIResult
 * @property {string} scanId
 * @property {string} timestamp
 * @property {string} modelVersion - e.g. "YOLOv8-AgriVision-v2.4 (Edge Quantized)"
 * @property {number} inferenceLatencyMs - e.g. 42ms
 * @property {number} confidence
 * @property {string} detectedLabel
 * @property {string} boundingBox - [x, y, width, height]
 * @property {string} status - 'Healthy' | 'Issue Detected'
 */
